### ISSUE TITLE
Figure out Hipchat API code

### ISSUE THREAD

**Issue by leftiness** [leftiness/lingo#7]
Haven't done the Hipchat API stuff yet.

Hipchat has a polling API. I'll have a thread running in the background asking Hipchat if there are new messages. When it finds some, it'll send a message to any registered listeners.

The state machine would be a registered listener if it is in a listening state. Then it'll handle updating the UI etc from there.

So the chat and the API polling are separate.

**Comment by leftiness** [comment]
Maybe I should separate the hipchat part into a truly separate application? It could run on a server and provide a web socket API for messages? Just a thought.

**Comment by leftiness** [comment]
If I make the hipchat polling code a separate process which sends messages to lingo over a socket, then lingo could be a multi protocol chat app. Maybe use a scripting language to make the hipchat part? Hmm...

**Comment by leftiness** [comment]
- `(trusty)brandon@localhost:~/secrets/lingo$ http get api.hipchat.com/v2/room 'Authorization: Bearer token_here'`
- `(trusty)brandon@localhost:~/secrets/lingo$ http post api.hipchat.com/v2/room/2896874/message 'Authorization: Bearer token_here' message='rest api message'`

**Comment by leftiness** [comment]
They have hooks now? Maybe they had them before and I didn't see the docs.

https://www.hipchat.com/docs/apiv2/webhooks

TLDR: I can register a URL and some configurations, and they'll send HTTP requests to that URL when events happen... user sends a message, etc.

This means I'll have to run an HTTP client and server for lingo. Not a big deal. I think I'll use Hyper.

**Comment by leftiness** [comment]
I looked a bit for hipchat libraries for rust. There's this one, but it's really just a wrapper around hyper, and I think I'd rather handle that myself. They do have some links to useful hipchat docs, though. http://rsolomo.github.io/hipchat-client/hipchat_client/client/struct.Client.html

**Comment by leftiness** [comment]
Okay. I think this issue might be closed. I'll open up more issues to implement other aspects of the hipchat api code that I need. For now, I've got something that will load rooms, decode them into rust structs, and store them on the state.
